Show a loading UI when accepting a dispute is in progress

Jinksi opened the pull request:
Fixes #7473

### Changes proposed in this Pull Request

This PR improves the "Accept dispute" merchant flow by showing a loading state within the Accept dispute modal when the accept dispute fetch request is in progress.

**Before**:
* the merchant clicks "accept dispute"
* the modal closes
* there is no indication of a network request in progress
* on success: a success snackbar message appears and the UI is updated with new dispute details
* on error: an error snackbar message appears

**After**:
* the merchant clicks "accept dispute"
* the modal remains open with a loading animation
* buttons are disabled and the modal cannot be closed
* on success: the modal closes, a success snackbar message appears and the UI is updated with new dispute details
* on error: the modal remains open, buttons are re-enabled and an error snackbar message appears

**TODO**
- [x] Change button text to `Accepting...` when in progress.
- [x] Reduce modal max-width to match design


### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Dispute: Needs response**
* As a shopper, create a disputed order by purchasing a test product using the card `4000000000000259` at checkout.
* As a merchant, navigate to Payments -> Disputes and click the row for the dispute you just created. You will be taken to the "Transaction Details" screen.
* Click the **Accept dispute** button to view the modal.
	* Ensure the modal can be closed by pressing cancel, close or clicking outside.
* Error handling
	* Disable network (Network tab → throttling) in your browser dev tools
	* Click **Accept dispute** and ensure the error is handled with a relevant message. The UI buttons should be re-enabled after the error event.
* Loading state:
	* Re-enable network (Network tab → throttling) in your browser dev tools
	* Click **Accept dispute** and enjoy the new loading animation for a few moments. Buttons should be disabled while the network request is in progress. The modal should not be dismissible.
	* Ensure the "Transaction Details" screen is refreshed with the following details:
		* `Disputed: Lost status`.
		* A message beginning with `This dispute was accepted and lost on {date}`.

**Inquiry: Needs response**
* As a shopper, create a disputed order by purchasing a test product using the card `4000000000001976` at checkout.
* As a merchant, navigate to Payments -> Disputes and click the row for the dispute you just created. You will be taken to the "Transaction Details" screen.
* Click **Issue refund** to view the modal.
	* Ensure the modal can be closed by pressing cancel, close or clicking outside.
* Ensure clicking **View order to issue refund** takes you to the "Edit order" screen.

### Screen recordings

**Screen recording of before without loading state**

https://github.com/Automattic/woocommerce-payments/assets/3147296/dcfbdea8-256d-445e-9c29-ebdec7defaad


**Screen recording of happy path** (with a slow network to demonstrate loading state)

https://github.com/Automattic/woocommerce-payments/assets/3147296/d620fa79-2760-4516-a85d-372484bd366f

**Screen recording of network error**

https://github.com/Automattic/woocommerce-payments/assets/3147296/700764a7-def4-4b67-a9eb-0e09b9e1ff60

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Jinksi committed:
Disable accept dispute modal close when request in progress

Jinksi committed:
Disable accept dispute modal buttons when request in progress

Jinksi committed:
Show "Accept dispute" button loading animation when in progress

Jinksi committed:
Leave accept dispute modal open when clicking accept or issue inquiry refund

Jinksi committed:
Add changelog entry

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7476` or branch name `fix/7473-accept-dispute-loading-state-ui` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/7473-accept-dispute-loading-state-ui) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 3e8f9ca91a3efb532b40a37e21e0230453def2d3
- Build time: 2023-10-25 03:39:02 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Jinksi commented:
FYI, I've requested a review but am yet to change the button text to `Accepting...` when in progress.

Feel free to review aside from this change.

Jinksi commented:
@nikkivias I've requested a review from you if you want to take a look. It is probably easiest to review the screen recordings I've added.

FYI, I am yet to change the button text to `Accepting...` when in progress.

nikivias commented:
@Jinksi The transition has improved significantly as it no longer jumps into the next screen.

I'm noticing other parts that don't quite match the Figma design though. Particularly the accept button which is outlined with only little side paddings (instead of Text only version), and the size of the modal is quite wide.

<img width="187" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/99638528/fd75d06f-63de-4f70-9224-cefe66ca4bb6">
<img width="542" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/99638528/844f14f4-aceb-4526-a42b-df1588c2e567">

<img width="647" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/99638528/5569edd5-1469-4566-93d1-b4129cf58768">
<img width="670" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/99638528/f629b01b-716c-4154-8ca1-c935e770f552">

Jinksi committed:
Merge branch 'develop' into fix/7473-accept-dispute-loading-state-ui

Jinksi committed:
Merge branch 'develop' into fix/7473-accept-dispute-loading-state-ui

Jinksi committed:
Fix accept dispute modal max-width

Jinksi commented:
Thanks for your feedback @nikkivias.

> the accept button which is outlined with only little side paddings (instead of Text only version)

This button uses the text-only `tertiary` button and I think this matches the component used in the Figma design. It will only show the border when hovered or focused.

https://github.com/Automattic/woocommerce-payments/assets/3147296/68040e68-3bff-4f5d-a17c-da134194f67e


> and the size of the modal is quite wide.

✅ I've fixed the max-width of the modal to match the design (600px).

<img width="682" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/0d823421-2c4d-4b41-8bba-4d32cea96a2f">

✅ I've also updated the button label to `Accepting…` when the request is in progress.

Jinksi committed:
Use `Accepting` button label when in progress

nikivias commented:
Thank you for actioning on those last 2 @Jinksi everything looks good!

Jinksi committed:
Merge branch 'develop' into fix/7473-accept-dispute-loading-state-ui

Jinksi commented:
Thanks @nikkivias! Requesting a code review from @Automattic/helix before merging.

haszari commented on client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:
```diff
@@ -82,10 +82,14 @@ interface AcceptDisputeProps {
+}: {
+	dispute: Dispute;
+	isLoading: boolean;
+} ): AcceptDisputeProps {
```
`getAcceptDisputeProps` is pretty strange to read and reason about. I think this code would be a lot clearer if it just had a dedicated modal for each case (inquiry vs dispute). `AcceptDisputeProps` is a domain-specific interface for a semi-generic modal (e.g. `modalLines`); I don't see a benefit to that over using the React components we have (i.e. a powerful, readable, flexible modal API). 

That said, I don't know if it's safe/worthwhile to refactor this for this change. I would be tempted 😁 

Mentioning because this PR reads as very decoupled – can't really see what changed. Specifically, the button `Submitting…` label should be an implementation detail inside a modal component. I find it strange (and complex) that this is delegated to a function (that now needs a new param).

Jinksi replied:
Yes, I agree and I began a refactor of this code in #7339, but it became low-priority and out-of-scope due to time constraints.

I think a refactor should be handled in a separate PR, as it is a more significant quantity of work and is separate from this PR's goal.

haszari commented on client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:
```diff
@@ -146,7 +150,9 @@ function getAcceptDisputeProps( dispute: Dispute ): AcceptDisputeProps {
 			},
 		],
-		modalButtonLabel: __( 'Accept dispute', 'woocommerce-payments' ),
+		modalButtonLabel: isLoading
```
`isLoading` could provide more info about what is loading. I think renaming is worthwhile because `isLoading` sounds like a pre-display state, not a "submitting" state. I realise this is an existing name in parent component, I think it's useful to rename there to be clear about what is loading.

The longest, least ambiguous name I can think of is `isAcceptDisputeSubmitPending` – that might be too much 😁

Jinksi replied:
Renamed to `isDisputeAcceptRequestPending` in b60c8dc.

haszari commented on client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:
```diff
@@ -335,6 +345,8 @@ const DisputeAwaitingResponseDetails: React.FC< Props > = ( {
 										<Button
 											variant="primary"
+											isBusy={ isLoading }
+											disabled={ isLoading }
```
Renaming `isLoading` would make this clearer IMO. E.g. 

```jsx
<Button
  label="submit"
  disabled={ isSubmitting }
/>
```

Jinksi replied:
Renamed to `isDisputeAcceptRequestPending` in b60c8dc.

haszari commented on client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:
```diff
@@ -346,7 +358,7 @@ const DisputeAwaitingResponseDetails: React.FC< Props > = ( {
 															'transaction_details',
 													}
 												);
-												setModalOpen( false );
```
Curious why we didn't use `onModalClose` before.

Jinksi replied:
Same! I don't have an answer.

haszari commented on client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:
```diff
@@ -171,6 +177,9 @@ const DisputeAwaitingResponseDetails: React.FC< Props > = ( {
 	const onModalClose = () => {
+		if ( isLoading ) {
+			return;
+		}
```
More renaming suggestions 😁 

`onModalClose` reads strange to me – what is its purpose? Currently the name is based on where it is bound (a kind of coupling?). With the addition of the early return, it's becoming abstract, e.g. `onModalClose() { if isLoading don't close }`.

Jinksi replied:
I've renamed it to `handleModalClose` in 27e7621 – does this make more sense?

haszari replied:
A tiny bit!

haszari reviewed (approved):
Tested and the new UI and submitting progress state is a vast improvement 🚀

I added some comments about renaming and/or refactoring the code to simplify (TLDR – make the code a closer match for what the merchant experiences). None of this is blocking, though renaming `isLoading` is probably safe and impactful.

Jinksi committed:
Merge branch 'develop' into fix/7473-accept-dispute-loading-state-ui

Jinksi committed:
Rename isLoading → isDisputeAcceptRequestPending

Jinksi committed:
Rename onModalClose → handleModalClose

Jinksi committed:
Merge branch 'develop' into fix/7473-accept-dispute-loading-state-ui
